modules/renter/download.go

```diff
@@ -208,94 +310,14 @@ func (r *Renter) Download(path, destination string) error {
-	progressDeadline := time.Now().Add(120 * time.Minute)
-	for {
-		// copy file contracts
-		file.mu.RLock()
-		contracts := make([]fileContract, 0, len(file.contracts))
-		for _, c := range file.contracts {
-			contracts = append(contracts, c)
-		}
-		file.mu.RUnlock()
-		if len(contracts) < file.erasureCode.MinPieces() {
-			return fmt.Errorf("contracts could not be located for this file - file may not be recoverable - needed %v, got %v", file.erasureCode.MinPieces(), len(contracts))
-		}
-
-		// interrupt upload loop
-		lockID = r.mu.Lock()
-		r.downloading = true
-		uploading := r.uploading
-		r.mu.Unlock(lockID)
-		resumeUploads := func() {
-			lockID = r.mu.Lock()
-			r.downloading = false
-			r.mu.Unlock(lockID)
-		}
-		// wait up to 60 minutes for upload loop to exit
-		timeout := time.Now().Add(60 * time.Minute)
-		for uploading && time.Now().Before(timeout) {
-			time.Sleep(time.Second)
-			lockID = r.mu.RLock()
-			uploading = r.uploading
-			r.mu.RUnlock(lockID)
-		}
-		if uploading {
-			resumeUploads()
-			return errors.New("timed out waiting for uploads to finish")
-		}
+	// Create the download object and add it to the queue.
+	d := file.newDownload(destination)
+	lockID = r.mu.Lock()
+	r.downloadQueue = append(r.downloadQueue, d)
+	r.mu.Unlock(lockID)
 
-		// Grab a set of hosts and attempt a download.
-		done, err := func() (bool, error) {
-			// Initiate connections to each host.
-			var hosts []fetcher
-			var errs []string
-			for _, c := range file.contracts {
-				d, err := r.hostContractor.Downloader(c.ID)
-				if err != nil {
-					errs = append(errs, fmt.Sprintf("\t%v: %v", c.IP, err))
-					continue
-				}
-				defer d.Close()
-				hosts = append(hosts, newHostFetcher(d, c.Pieces, file.masterKey))
-			}
-			if len(hosts) < file.erasureCode.MinPieces() {
-				return false, errors.New("could not connect to enough hosts:\n" + strings.Join(errs, "\n"))
-			}
-			// Check that this host set is sufficient to download the file.
-			err := checkHosts(hosts, file.erasureCode.MinPieces(), file.numChunks())
-			if err != nil {
-				return false, err
-			}
-			// Update the downloader with the new set of hosts.
-			d.hosts = hosts
-
-			// Perform download.
-			err = d.run(f)
-			done := err == nil
-			return done, nil
-		}()
-		if done {
-			// Download is complete!
-			resumeUploads()
-			break
-		} else if err != nil {
-			// One of the more severe errors occurred, wait a bit before trying
-			// the download again.
-			resumeUploads()
-			time.Sleep(time.Second * 90)
-		} else {
-			// We made progress, but haven't finished yet. Reset the progress
-			// deadline.
-			progressDeadline = time.Now().Add(30 * time.Minute)
-		}
-		// if we haven't made any progress in 30 minutes, give up
-		if time.Now().After(progressDeadline) {
-			return errors.New("no progress in 30 minutes; giving up")
-		}
-	}
-	return nil
+	// Block until the download has completed.
+	return <-d.downloadFinished
```

Comment:
better to just return `d.downloadFinished`; that makes it dead simple for the API to add a non-blocking version of the `/download` route